File: mobile/lib/models/spam_message.dart
class SpamMessage {
  final int? id;
  final String userId;
  final String phoneNumber;
  final String messageText;
  final String detectionMethod; // 'ml', 'ai', 'both'
  final String threatLevel; // 'low', 'medium', 'high'
  final double? aiConfidence;
  final double? mlConfidence;
  final DateTime detectedAt;
  final bool isRead;
  final bool isFalsePositive;

  SpamMessage({
    this.id,
    required this.userId,
    required this.phoneNumber,
    required this.messageText,
    required this.detectionMethod,
    required this.threatLevel,
    this.aiConfidence,
    this.mlConfidence,
    required this.detectedAt,
    this.isRead = false,
    this.isFalsePositive = false,
  });

  factory SpamMessage.fromJson(Map<String, dynamic> json) {
    return SpamMessage(
      id: json['id'],
      userId: json['user_id'],
      phoneNumber: json['phone_number'],
      messageText: json['message_text'],
      detectionMethod: json['detection_method'],
      threatLevel: json['threat_level'],
      aiConfidence: json['ai_confidence']?.toDouble(),
      mlConfidence: json['ml_confidence']?.toDouble(),
      detectedAt: DateTime.parse(json['detected_at']),
      isRead: json['is_read'] ?? false,
      isFalsePositive: json['is_false_positive'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'phone_number': phoneNumber,
      'message_text': messageText,
      'detection_method': detectionMethod,
      'threat_level': threatLevel,
      'ai_confidence': aiConfidence,
      'ml_confidence': mlConfidence,
      'detected_at': detectedAt.toIso8601String(),
      'is_read': isRead,
      'is_false_positive': isFalsePositive,
    };
  }
}

class SecurityStats {
  final int total;
  final int unread;
  final int highThreat;
  final int mediumThreat;
  final int lowThreat;
  final int today;
  final int thisWeek;

  SecurityStats({
    required this.total,
    required this.unread,
    required this.highThreat,
    required this.mediumThreat,
    required this.lowThreat,
    required this.today,
    required this.thisWeek,
  });

  factory SecurityStats.fromJson(Map<String, dynamic> json) {
    return SecurityStats(
      total: int.parse(json['total'].toString()),
      unread: int.parse(json['unread'].toString()),
      highThreat: int.parse(json['high_threat'].toString()),
      mediumThreat: int.parse(json['medium_threat'].toString()),
      lowThreat: int.parse(json['low_threat'].toString()),
      today: int.parse(json['today'].toString()),
      thisWeek: int.parse(json['this_week'].toString()),
    );
  }
}

File: backend/models/SpamMessage.js
const pool = require('../config/db');

class SpamMessage {
    static async create({
        user_id,
        phone_number,
        message_text,
        detection_method,
        threat_level,
        ai_confidence = null,
        ml_confidence = null
    }) {
        const query = `
            INSERT INTO spam_messages (
                user_id, phone_number, message_text, detection_method,
                threat_level, ai_confidence, ml_confidence
            )
            VALUES ($1, $2, $3, $4, $5, $6, $7)
            RETURNING *;
        `;
        const values = [
            user_id, phone_number, message_text, detection_method,
            threat_level, ai_confidence, ml_confidence
        ];
        const { rows } = await pool.query(query, values);
        return rows[0];
    }

    static async findByUserId(user_id, { limit = 50, offset = 0, unreadOnly = false } = {}) {
        let query = `
            SELECT * FROM spam_messages
            WHERE user_id = $1
        `;

        if (unreadOnly) {
            query += ' AND is_read = false';
        }

        query += `
            ORDER BY detected_at DESC
            LIMIT $2 OFFSET $3;
        `;

        const { rows } = await pool.query(query, [user_id, limit, offset]);
        return rows;
    }

    static async findById(id, user_id) {
        const query = `
            SELECT * FROM spam_messages
            WHERE id = $1 AND user_id = $2;
        `;
        const { rows } = await pool.query(query, [id, user_id]);
        return rows[0];
    }

    static async markAsRead(id, user_id) {
        const query = `
            UPDATE spam_messages
            SET is_read = true
            WHERE id = $1 AND user_id = $2
            RETURNING *;
        `;
        const { rows } = await pool.query(query, [id, user_id]);
        return rows[0];
    }

    static async markAsFalsePositive(id, user_id) {
        const query = `
            UPDATE spam_messages
            SET is_false_positive = true
            WHERE id = $1 AND user_id = $2
            RETURNING *;
        `;
        const { rows } = await pool.query(query, [id, user_id]);
        return rows[0];
    }

    static async getStats(user_id) {
        const query = `
            SELECT
                COUNT(*) as total,
                COUNT(*) FILTER (WHERE is_read = false) as unread,
                COUNT(*) FILTER (WHERE threat_level = 'high') as high_threat,
                COUNT(*) FILTER (WHERE threat_level = 'medium') as medium_threat,
                COUNT(*) FILTER (WHERE threat_level = 'low') as low_threat,
                COUNT(*) FILTER (WHERE detected_at >= CURRENT_DATE) as today,
                COUNT(*) FILTER (WHERE detected_at >= CURRENT_DATE - INTERVAL '7 days') as this_week
            FROM spam_messages
            WHERE user_id = $1;
        `;
        const { rows } = await pool.query(query, [user_id]);
        return rows[0];
    }

    static async deleteById(id, user_id) {
        const query = `
            DELETE FROM spam_messages
            WHERE id = $1 AND user_id = $2
            RETURNING id;
        `;
        const { rows } = await pool.query(query, [id, user_id]);
        return rows[0];
    }
}

module.exports = SpamMessage;
